README.md
### **Named Entity Recognition for extracting Open Source Hardware project metadata**

- In this repository you can find an application, which introduces a main focus in Named Entity Recognition (NLP or NER).
- The aim is to extract common entities within a text (or webpages).
- For this purpose, the Spacy library is used to train a deep learning model based on neural netrworks to detect entities from text data.
- To be able to train the model, it will also show how to create a train dataset and label them in order to perform NER.
- This project is designed for the end users or developers, who need to receive the charateristics of OSH projects. 
- The selected characteristics are manufacturing process, machine type, material and dimensions. 
- As a future step, this informations can be linked to the suitable manufacturers for users to find corresponding Makerspace/Fablabs in order to manufacture their prototypes or individualized products.

## **CURRENT STATUS**

- Technology status of the project is [OTRL 6](https://github.com/OPEN-NEXT/LOSH/blob/master/OTRL.md#concept), which means it is a ready-to-use product and demonstrated in relevant environment.
- Documentation status of the project is [ODRL 3](https://github.com/OPEN-NEXT/LOSH/blob/master/OTRL.md#concept), which means it is an early release.
- The next steps are
  - Completing the documentation
  - New use cases
  - Enhancing the algorithm for a better accuracy
- The needed skills are basic python skills and some enthasuiasm.

## **THE PROBLEM**

- There were already a lot of algortihms for Named Entity Recognition, but not for this specific topic with manufacturing characteristics. The algortihm can still be trained with other train data or for other parameters to extract other needed information and enhance the use cases.

## **PRODUCT FEATURES/FUNCTIONS**

- The algorithm uses NER of SpaCy to train the model with Deep Learning (NN) for the characteristics mentioned. 
- With a web application any kind of text can be given as input, and the output could be received with the classified entities.
- The application can be used for any plain text input or Mediawiki-based websites (while this was only tested on Appropedia)

## **THE TEAM**

- As a research partner of the Open!Next project, a small team from Fraunhofer IPK has created the solution so as to contribute for OSH.
- OPEN!NEXT received funding from the European Union’s Horizon 2020 research and innovation programme under grant agreement no. 869984.
- Our team has some basic Python skills.

## **ABOUT THE DESIGN**

- We will perform the following: (how is this process with the application?)
  - Create a train dataset from OSH project websites. (Giving example?)
  - Label the dataset with the selected entities using Doccano labeling tool manually.
  - Save the labels in a text file as JSONL.
  - Spacy Neural Network model to train a new statistical model.
  - We will save the model.
  - We will create a Spacy NLP pipeline and use the new model to detect manufacturing entities.


## PROGRAMMING

- Needed softwares are
  - Python Compiler (Python 3.7 or more)
  - Doccano (1.3.0)
- Needed libraries are
  - SpaCy (3.0.5)
  - json 
  - Tkinter
  - requets
  - bs4 as Beautiful Soup
  - Pandas
  - Pickle
- Need to download an IDE environmennt on you computer, clone the repository on you IDE, then to run the application, run either runModel_support file one or two 


## **INSTRUCTIONS TO RUN THE APPLICATION**
- Copy the URL of the repository
- Open the python IDE which allows to clone github repository, in our case we used PyCharm Community Edition 2020.3.2
-	create a new project and save it
- In the project, go to vcs and click on create git repository (for the ide it may be that you need to install git)
- Instead of vcs, it will show git, under git select clone and paste the url
-	It will ask to login (when private)
-	After cloning the repository, it will open in your ide, you need to create the python environment if it has not happened automaticalyy, and the python compiler for opening application 1 or 2, go from directory to selected app, run the runModel_support 
- The application will start in a new window, you can give your input and see the results

## **INSTRUCTIONS TO MODIFY**

### CREATING A DATASET
- To create a dataset, our team used basic definitions and some open source hardware plattforms. The train dataset is saved as txt to import in doccano later on.

### LABELING THE DATA
-This step is for labeling the entities using Doccano, but if you already have labeled data, you can skip this step and directly go to training the model.
- First step is to install doccano, please follow the [doccano](https://doccano.github.io/doccano/getting-started/) instructions and open the program
- For WIndows installation,
  - `pip install doccano`
  - `doccano`
- Go to http://127.0.0.1:8000/
- Login with username: admin and password:password
<img src="https://user-images.githubusercontent.com/60435723/116371480-bfbeef00-a80b-11eb-8bc1-429b771d1ac5.png" width="400">
- Click create, type in your project name, description and select the sequence labeling project type
<img src="https://user-images.githubusercontent.com/60435723/116376637-bdab5f00-a810-11eb-817b-8e5f7cef8db9.png" width="400">
- After creating project, click on dataset an import your dataset
- Go to Labels and create your labels
- Go back to your imported dataset and click annotate
- The first entity will open, you can select the word/s you want to label and continue untill all your dataset is labeled accordingly
- After finishing the labeling process, click "Export Dataset" as JSONL file format under Actions and save the file

### TRAINING THE MODEL
- First step is to install spacy
- Firstly we read the JSONL file:
```
  import json
  labeled_data = []
  with open(r"project_1_dataset_v4.jsonl", "r", encoding='utf-8') as read_file:
    for line in read_file:
        data = json.loads(line)
        labeled_data.append(data)
  print(labeled_data)
```

- After reading our data, we need to convert the format
 ```
 TRAINING_DATA = []
  for entry in labeled_data:
      entities = []
      for e in entry['labels']:
          entities.append((e[0], e[1],e[2]))
      spacy_entry = (entry['text'], {"entities": entities})
      TRAINING_DATA.append(spacy_entry)
 print(TRAINING_DATA)
```
- Next step is to train the model- We use Deep Learning (NN) and set a dropout rate of 0.3 to prevent overfitting.
```
import spacy
import random
import json
from spacy.tokens import Doc
from spacy.training import Example
nlp = spacy.blank("en")
ner = nlp.create_pipe("ner")
nlp.add_pipe('ner')
for _, annotations in TRAINING_DATA:                    #goes through all the entities are get the name token.label_ one
    for ent in annotations.get("entities"):
        ner.add_label(ent[2])
# Start the training
nlp.begin_training()
# Loop for 40 iterations
for itn in range(40):
    # Shuffle the training data
    random.shuffle(TRAINING_DATA)
    losses = {}
# Batch the examples and iterate over them
    for batch in spacy.util.minibatch(TRAINING_DATA, size=2):
        for text, annotations in batch:
        # create Example
            doc = nlp.make_doc(text)
            example = Example.from_dict(doc, annotations)
            # Update the model
            nlp.update([example], losses=losses, drop=0.3)
            example = Example.from_dict(doc, annotations)
    print(losses)
```
- While training data, you can receive some warnings at first, then the iterations should start and take a few minutes
<img src="https://user-images.githubusercontent.com/60435723/116433904-3a0e6400-a84a-11eb-939c-7070aaa4d3c0.png" width="400">

- After iterations stop, we save the model

```
 nlp.to_disk("./my.model")
```
- Testing the model

```
from spacy import displacy
example = "an example test"
doc = nlp(example)
displacy.render(doc, style='ent')
```

## **POTENTIAL IMPROVEMENTS**
- The algorithm can be used in different use cases with corresponding train data


References used:
1. [Reference 1](https://itnext.io/nlp-named-entity-recognition-ner-with-spacy-and-python-dabaf843cab2?gi=493e297bfed)


Raw Data/History.txt
is the study of the past.Events occurring before the invention of writing systems are considered prehistory. "History" is an umbrella term that relates to past events as well as the memory, discovery, collection, organization, presentation, and interpretation of information about these events. Historians place the past in context using historical sources such as written documents, oral accounts, ecological markers, and material objects including art and artifacts.
History also includes the academic discipline which uses narrative to describe, examine, question, and analyze a sequence of past events, and investigate the patterns of cause and effect that are related to them.
Historians seek to understand and represent the past through narratives. They often debate which narrative best explains an event, as well as the significance of different causes and effects. Historians also debate the nature of history and its usefulness by discussing the study of the discipline as an end in itself and as a way of providing "perspective" on the problems of the present
Stories common to a particular culture, but not supported by external sources (such as the tales surrounding King Arthur), are usually classified as cultural heritage or legends
History differs from myth in that it is supported by evidence. However, ancient influences have helped spawn variant interpretations of the nature of history which have evolved over the centuries and continue to change today. The modern study of history is wide-ranging, and includes the study of specific regions and the study of certain topical or thematic elements of historical investigation. History is often taught as part of primary and secondary education, and the academic study of history is a major discipline in university studies.
Herodotus, a 5th-century BC Greek historian is often considered (within the Western tradition) to be the "father of history", or, the "father of lies". 
Along with his contemporary Thucydides, he helped form the foundations for the modern study of human history. Their works continue to be read today, and the gap between the culture-focused Herodotus and the military-focused Thucydides remains a point of contention or approach in modern historical writing. In East Asia, a state chronicle, the Spring and Autumn Annals, was known to be compiled from as early as 722 BC although only 2nd-century BC texts have survived.


Model/NERapp/runModel.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# GUI module generated by PAGE version 6.0.1
#  in conjunction with Tcl version 8.6
#    Apr 09, 2021 03:30:35 PM CEST  platform: Windows NT

import sys

try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True

import runModel_support

def vp_start_gui():
    '''Starting point when module is the main routine.'''
    global val, w, root
    root = tk.Tk()
    runModel_support.set_Tk_var()
    top = Toplevel1 (root)
    runModel_support.init(root, top)
    root.mainloop()

w = None
def create_Toplevel1(rt, *args, **kwargs):
    '''Starting point when module is imported by another module.
       Correct form of call: 'create_Toplevel1(root, *args, **kwargs)' .'''
    global w, w_win, root
    #rt = root
    root = rt
    w = tk.Toplevel (root)
    runModel_support.set_Tk_var()
    top = Toplevel1 (w)
    runModel_support.init(w, top, *args, **kwargs)
    return (w, top)

def destroy_Toplevel1():
    global w
    w.destroy()
    w = None

class Toplevel1:
    def __init__(self, top=None):
        '''This class configures and populates the toplevel window.
           top is the toplevel containing window.'''
        _bgcolor = '#d9d9d9'  # X11 color: 'gray85'
        _fgcolor = '#000000'  # X11 color: 'black'
        _compcolor = '#d9d9d9' # X11 color: 'gray85'
        _ana1color = '#d9d9d9' # X11 color: 'gray85'
        _ana2color = '#ececec' # Closest X11 color: 'gray92'

        top.geometry("600x450+660+210")
        top.minsize(148, 1)
        top.maxsize(1924, 1055)
        top.resizable(1,  1)
        top.title("NER Application")
        top.configure(background="#d9d9d9")

        self.Button1 = tk.Button(top)
        self.Button1.place(relx=0.45, rely=0.533, height=33, width=166)
        self.Button1.configure(activebackground="#ececec")
        self.Button1.configure(activeforeground="#000000")
        self.Button1.configure(background="#d9d9d9")
        self.Button1.configure(command=runModel_support.runButton)
        self.Button1.configure(disabledforeground="#a3a3a3")
        self.Button1.configure(foreground="#000000")
        self.Button1.configure(highlightbackground="#d9d9d9")
        self.Button1.configure(highlightcolor="black")
        self.Button1.configure(pady="0")
        self.Button1.configure(text='''Run Model''')

        self.Label1 = tk.Label(top)
        self.Label1.place(relx=0.2, rely=0.289, height=26, width=172)
        self.Label1.configure(background="#d9d9d9")
        self.Label1.configure(disabledforeground="#a3a3a3")
        self.Label1.configure(foreground="#000000")
        self.Label1.configure(text='''Enter Corpus txt''')

        self.Entry1 = tk.Entry(top)
        self.Entry1.place(relx=0.533, rely=0.289, height=24, relwidth=0.357)
        self.Entry1.configure(background="white")
        self.Entry1.configure(disabledforeground="#a3a3a3")
        self.Entry1.configure(font="TkFixedFont")
        self.Entry1.configure(foreground="#000000")
        self.Entry1.configure(insertbackground="black")
        self.Entry1.configure(textvariable=runModel_support.corpus)

if __name__ == '__main__':
    vp_start_gui()







Model/NERapp2/runModel2.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# GUI module generated by PAGE version 6.0.1
#  in conjunction with Tcl version 8.6
#    Apr 09, 2021 03:30:35 PM CEST  platform: Windows NT

import sys

try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True

import runModel_support2

def vp_start_gui():
    '''Starting point when module is the main routine.'''
    global val, w, root
    root = tk.Tk()
    runModel_support2.set_Tk_var()
    top = Toplevel1 (root)
    runModel_support2.init(root, top)
    root.mainloop()

w = None
def create_Toplevel1(rt, *args, **kwargs):
    '''Starting point when module is imported by another module.
       Correct form of call: 'create_Toplevel1(root, *args, **kwargs)' .'''
    global w, w_win, root
    #rt = root
    root = rt
    w = tk.Toplevel (root)
    runModel_support2.set_Tk_var()
    top = Toplevel1 (w)
    runModel_support2.init(w, top, *args, **kwargs)
    return (w, top)

def destroy_Toplevel1():
    global w
    w.destroy()
    w = None

class Toplevel1:
    def __init__(self, top=None):
        '''This class configures and populates the toplevel window.
           top is the toplevel containing window.'''
        _bgcolor = '#d9d9d9'  # X11 color: 'gray85'
        _fgcolor = '#000000'  # X11 color: 'black'
        _compcolor = '#d9d9d9' # X11 color: 'gray85'
        _ana1color = '#d9d9d9' # X11 color: 'gray85'
        _ana2color = '#ececec' # Closest X11 color: 'gray92'

        top.geometry("600x450+660+210")
        top.minsize(148, 1)
        top.maxsize(1924, 1055)
        top.resizable(1,  1)
        top.title("NER Application 2")
        top.configure(background="#d9d9d9")

        self.Button1 = tk.Button(top)
        self.Button1.place(relx=0.45, rely=0.533, height=33, width=166)
        self.Button1.configure(activebackground="#ececec")
        self.Button1.configure(activeforeground="#000000")
        self.Button1.configure(background="#d9d9d9")
        self.Button1.configure(command=runModel_support2.runButton)
        self.Button1.configure(disabledforeground="#a3a3a3")
        self.Button1.configure(foreground="#000000")
        self.Button1.configure(highlightbackground="#d9d9d9")
        self.Button1.configure(highlightcolor="black")
        self.Button1.configure(pady="0")
        self.Button1.configure(text='''Run Model''')

        self.Label1 = tk.Label(top)
        self.Label1.place(relx=0.2, rely=0.289, height=26, width=172)
        self.Label1.configure(background="#d9d9d9")
        self.Label1.configure(disabledforeground="#a3a3a3")
        self.Label1.configure(foreground="#000000")
        self.Label1.configure(text='''Enter Appropedia URL''')

        self.Entry1 = tk.Entry(top)
        self.Entry1.place(relx=0.533, rely=0.289, height=24, relwidth=0.357)
        self.Entry1.configure(background="white")
        self.Entry1.configure(disabledforeground="#a3a3a3")
        self.Entry1.configure(font="TkFixedFont")
        self.Entry1.configure(foreground="#000000")
        self.Entry1.configure(insertbackground="black")
        self.Entry1.configure(textvariable=runModel_support2.corpus)

if __name__ == '__main__':
    vp_start_gui()







Model/NERapp/runModel_support.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# Support module generated by PAGE version 6.0.1
#  in conjunction with Tcl version 8.6
#    Apr 09, 2021 03:30:41 PM CEST  platform: Windows NT

import sys
from spacy import displacy
import spacy
import os


try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True

def set_Tk_var():
    global corpus
    corpus = tk.StringVar()

def init(top, gui, *args, **kwargs):
    global w, top_level, root
    w = gui
    top_level = top
    root = top

def runButton():
    example = corpus.get()
    print("I got the corpus")
    nlp = spacy.load("./best_model")
    print("NER loaded")
    doc = nlp(example)

    html = displacy.render(doc, style='ent')
    with open("data_visualisation.html", "w") as file:
        file.write(html)
    print("html page created")
    file.close()

    import webbrowser
    new = 2
    url = os.getcwd() + '/data_visualisation.html'
    print(url)
    webbrowser.open(url, new=new)
    print("html page opened")
    sys.stdout.flush()

def destroy_window():
    # Function which closes the window.
    global top_level
    top_level.destroy()
    top_level = None

if __name__ == '__main__':
    import runModel
    runModel.vp_start_gui()






Model/NERapp2/runModel_support2.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# Support module generated by PAGE version 6.0.1
#  in conjunction with Tcl version 8.6
#    Apr 09, 2021 03:30:41 PM CEST  platform: Windows NT

import sys
from spacy import displacy
import spacy
import requests
from bs4 import BeautifulSoup
import pandas as pd
import pickle
import os

try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True


def url_to_transcript(url):
    '''Returns transcript data specifically from scrapsfromtheloft.com.'''
    page = requests.get(url).text
    soup = BeautifulSoup(page, "lxml")
    text = [p.text for p in soup.find(class_="mw-content-ltr")]
    #print(url)
    return text

def set_Tk_var():
    global corpus
    corpus = tk.StringVar()

def init(top, gui, *args, **kwargs):
    global w, top_level, root
    w = gui
    top_level = top
    root = top

def runButton():
    url = corpus.get()
    urls = []
    urls.append(url)
    transcripts = [url_to_transcript(u) for u in urls]
    hardwares = ["hardware"]

    for i, c in enumerate(hardwares):  # creating the txt file to load the data
        with open(c + ".txt", "wb") as file:
            pickle.dump(transcripts[i], file)

    data = {}
    for i, c in enumerate(hardwares):
        with open(c + ".txt", "rb") as file:
            data[c] = pickle.load(file)

    def combine_text(list_of_text):
        '''Takes a list of text and combines them into one large chunk of text.'''
        combined_text = ' '.join(list_of_text)
        return combined_text

    data_combined = {key: [combine_text(value)] for (key, value) in data.items()}

    pd.set_option('max_colwidth', 150)
    data_df = pd.DataFrame.from_dict(data_combined).transpose()
    data_df.columns = ['transcript']
    data_df = data_df.sort_index()

    import re
    import string

    def clean_text_round1(text):
        '''Make text lowercase, remove text in square brackets, remove punctuation and remove words containing numbers.'''
        text = text.lower()
        text = re.sub('\[.*?\]', '', text)
        text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
        return text

    round1 = lambda x: clean_text_round1(x)

    data_clean = pd.DataFrame(data_df.transcript.apply(round1))

    def clean_text_round2(text):
        '''Get rid of some additional punctuation and non-sensical text that was missed the first time around.'''
        text = re.sub('[‘’“”…]', '', text)
        text = re.sub('\n', ' ', text)
        # text = re.sub('\w*\d\w*', '', text)
        return text

    round2 = lambda x: clean_text_round2(x)

    data_clean = pd.DataFrame(data_clean.transcript.apply(round2))
    # print(data_clean)

    cleaned_transcripts = [data_clean.transcript.loc[i] for i in hardwares]

    for i, c in enumerate(hardwares):
        with open(c + ".txt", "w") as file:
            file.write(cleaned_transcripts[i])

    for i, c in enumerate(hardwares):
        with open(c + ".txt", "r") as file:
            example = file.read()

    print("I got the corpus")
    nlp = spacy.load("./best_model")
    print("NER loaded")
    doc = nlp(example)

    html = displacy.render(doc, style='ent')
    with open("data_visualisation2.html", "w") as file:
        file.write(html)
    print("html page created")
    file.close()

    import webbrowser
    new = 2
    url = os.getcwd() + '/data_visualisation2.html'
    webbrowser.open(url,new=new)
    print("html page opened")
    sys.stdout.flush()

def destroy_window():
    # Function which closes the window.
    global top_level
    top_level.destroy()
    top_level = None

if __name__ == '__main__':
    import runModel2
    runModel2.vp_start_gui()






Model/NERapp3/runModel3_support.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# Support module generated by PAGE version 6.0.1
#  in conjunction with Tcl version 8.6
#    Apr 28, 2021 02:08:10 PM CEST  platform: Windows NT

import sys
from spacy import displacy
import spacy
import requests
from bs4 import BeautifulSoup
import pandas as pd
import pickle
import os

try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True

def url_to_transcript(url):
    '''Returns transcript data specifically from scrapsfromtheloft.com.'''
    page = requests.get(url).text
    soup = BeautifulSoup(page, "lxml")
    text = [p.text for p in soup.find(class_="mw-content-ltr")]
    #print(url)
    return text

def set_Tk_var():
    global urlLabel
    urlLabel = tk.StringVar()

def set_Tk_var2():
    global corpusLabel
    corpusLabel = tk.StringVar()

def init(top, gui, *args, **kwargs):
    global w, top_level, root
    w = gui
    top_level = top
    root = top

def corpusButton():
    example = corpusLabel.get()
    print("I got the corpus")
    nlp = spacy.load("./best_model")
    print("NER loaded")
    doc = nlp(example)

    html = displacy.render(doc, style='ent')
    with open("data_visualisation3.html", "w") as file:
        file.write(html)
    print("html page created")
    file.close()

    import webbrowser
    new = 2
    url = os.getcwd() + '/data_visualisation3.html'
    print(url)
    webbrowser.open(url, new=new)
    print("html page opened")
    sys.stdout.flush()

def urlButton():
    url = urlLabel.get()
    urls = []
    urls.append(url)
    transcripts = [url_to_transcript(u) for u in urls]
    hardwares = ["hardware"]

    for i, c in enumerate(hardwares):  # creating the txt file to load the data
        with open(c + ".txt", "wb") as file:
            pickle.dump(transcripts[i], file)

    data = {}
    for i, c in enumerate(hardwares):
        with open(c + ".txt", "rb") as file:
            data[c] = pickle.load(file)

    def combine_text(list_of_text):
        '''Takes a list of text and combines them into one large chunk of text.'''
        combined_text = ' '.join(list_of_text)
        return combined_text

    data_combined = {key: [combine_text(value)] for (key, value) in data.items()}

    pd.set_option('max_colwidth', 150)
    data_df = pd.DataFrame.from_dict(data_combined).transpose()
    data_df.columns = ['transcript']
    data_df = data_df.sort_index()

    import re
    import string

    def clean_text_round1(text):
        '''Make text lowercase, remove text in square brackets, remove punctuation and remove words containing numbers.'''
        text = text.lower()
        text = re.sub('\[.*?\]', '', text)
        text = re.sub('[%s]' % re.escape(string.punctuation), '', text)
        return text

    round1 = lambda x: clean_text_round1(x)

    data_clean = pd.DataFrame(data_df.transcript.apply(round1))

    def clean_text_round2(text):
        '''Get rid of some additional punctuation and non-sensical text that was missed the first time around.'''
        text = re.sub('[‘’“”…]', '', text)
        text = re.sub('\n', ' ', text)
        # text = re.sub('\w*\d\w*', '', text)
        return text

    round2 = lambda x: clean_text_round2(x)

    data_clean = pd.DataFrame(data_clean.transcript.apply(round2))
    # print(data_clean)

    cleaned_transcripts = [data_clean.transcript.loc[i] for i in hardwares]

    for i, c in enumerate(hardwares):
        with open(c + ".txt", "w") as file:
            file.write(cleaned_transcripts[i])

    for i, c in enumerate(hardwares):
        with open(c + ".txt", "r") as file:
            example = file.read()

    print("I got the corpus")
    nlp = spacy.load("./best_model")
    print("NER loaded")
    doc = nlp(example)

    html = displacy.render(doc, style='ent')
    with open("data_visualisation3.html", "w") as file:
        file.write(html)
    print("html page created")
    file.close()

    import webbrowser
    new = 2
    url = os.getcwd() + '/data_visualisation3.html'
    webbrowser.open(url, new=new)
    print("html page opened")
    sys.stdout.flush()

def destroy_window():
    # Function which closes the window.
    global top_level
    top_level.destroy()
    top_level = None

if __name__ == '__main__':
    import runModel3
    runModel3.vp_start_gui()






Model/NERapp3/runModel3.py
#! /usr/bin/env python
#  -*- coding: utf-8 -*-
#
# GUI module generated by PAGE version 6.0.1
#  in conjunction with Tcl version 8.6
#    Apr 28, 2021 02:08:02 PM CEST  platform: Windows NT

import sys
from PIL import Image, ImageTk

try:
    import Tkinter as tk
except ImportError:
    import tkinter as tk

try:
    import ttk
    py3 = False
except ImportError:
    import tkinter.ttk as ttk
    py3 = True

import runModel3_support

def vp_start_gui():
    '''Starting point when module is the main routine.'''
    global val, w, root
    root = tk.Tk()
    runModel3_support.set_Tk_var()
    runModel3_support.set_Tk_var2()

    image1 = Image.open("IPK.png")
    image1 = image1.resize((201, 100), Image.ANTIALIAS)
    test = ImageTk.PhotoImage(image1)
    label1 = tk.Label(image=test)
    label1.image = test
    # Position image
    label1.place(relx=0, rely=0)

    image2 = Image.open("openNext.png")
    image2 = image2.resize((246, 160), Image.ANTIALIAS)
    test = ImageTk.PhotoImage(image2)
    label2 = tk.Label(image=test)
    label2.image = test
    # Position image
    label2.place(relx=0.59, rely=0.64)

    top = Toplevel1(root)
    runModel3_support.init(root, top)
    root.mainloop()

w = None
def create_Toplevel1(rt, *args, **kwargs):
    '''Starting point when module is imported by another module.
       Correct form of call: 'create_Toplevel1(root, *args, **kwargs)' .'''
    global w, w_win, root
    #rt = root
    root = rt
    w = tk.Toplevel (root)
    runModel3_support.set_Tk_var()
    top = Toplevel1 (w)
    runModel3_support.init(w, top, *args, **kwargs)
    return (w, top)

def destroy_Toplevel1():
    global w
    w.destroy()
    w = None

class Toplevel1:
    def __init__(self, top=None):
        '''This class configures and populates the toplevel window.
           top is the toplevel containing window.'''
        _bgcolor = '#d9d9d9'  # X11 color: 'gray85'
        _fgcolor = '#000000'  # X11 color: 'black'
        _compcolor = '#d9d9d9' # X11 color: 'gray85'
        _ana1color = '#d9d9d9' # X11 color: 'gray85'
        _ana2color = '#ececec' # Closest X11 color: 'gray92'
        self.style = ttk.Style()
        if sys.platform == "win32":
            self.style.theme_use('winnative')
        self.style.configure('.',background=_bgcolor)
        self.style.configure('.',foreground=_fgcolor)
        self.style.map('.',background=
            [('selected', _compcolor), ('active',_ana2color)])

        top.geometry("600x450+602+194")
        top.minsize(148, 1)
        top.maxsize(1924, 1030)
        top.resizable(1,  1)
        top.title("NER Application")
        top.configure(background="#2ca971")

        self.Button1 = tk.Button(top)
        self.Button1.place(relx=0.05, rely=0.8, height=53, width=156)
        self.Button1.configure(activebackground="#ececec")
        self.Button1.configure(activeforeground="#000000")
        self.Button1.configure(background="#d9d9d9")
        self.Button1.configure(command=runModel3_support.urlButton)
        self.Button1.configure(disabledforeground="#a3a3a3")
        self.Button1.configure(foreground="#000000")
        self.Button1.configure(highlightbackground="#d9d9d9")
        self.Button1.configure(highlightcolor="black")
        self.Button1.configure(pady="0")
        self.Button1.configure(text='''Run Model''')

        self.Button3 = tk.Button(top)
        self.Button3.place(relx=0.45, rely=0.35, height=53, width=136)
        self.Button3.configure(activebackground="#ececec")
        self.Button3.configure(activeforeground="#000000")
        self.Button3.configure(background="#d9d9d9")
        self.Button3.configure(command=runModel3_support.corpusButton)
        self.Button3.configure(disabledforeground="#a3a3a3")
        self.Button3.configure(foreground="#000000")
        self.Button3.configure(highlightbackground="#d9d9d9")
        self.Button3.configure(highlightcolor="black")
        self.Button3.configure(pady="0")
        self.Button3.configure(text='''Run Model''')

        self.Entry3 = tk.Entry(top)
        self.Entry3.place(relx=0.05, rely=0.675, height=54, relwidth=0.5)
        self.Entry3.configure(background="white")
        self.Entry3.configure(disabledforeground="#a3a3a3")
        self.Entry3.configure(font="TkFixedFont")
        self.Entry3.configure(foreground="#000000")
        self.Entry3.configure(insertbackground="black")
        self.Entry3.configure(textvariable=runModel3_support.urlLabel)

        self.Entry4 = tk.Entry(top)
        self.Entry4.place(relx=0.45, rely=0.225, height=54, relwidth=0.5)
        self.Entry4.configure(background="white")
        self.Entry4.configure(disabledforeground="#a3a3a3")
        self.Entry4.configure(font="TkFixedFont")
        self.Entry4.configure(foreground="#000000")
        self.Entry4.configure(insertbackground="black")
        self.Entry4.configure(textvariable=runModel3_support.corpusLabel)

        self.Label1 = tk.Label(top)
        self.Label1.place(relx=0.45, rely=0.15, height=26, width=171)
        self.Label1.configure(background="#d9d9d9")
        self.Label1.configure(disabledforeground="#a3a3a3")
        self.Label1.configure(foreground="#000000")
        self.Label1.configure(text='''Enter a corpus text below''')

        self.Label2 = tk.Label(top)
        self.Label2.place(relx=0.05, rely=0.6, height=26, width=171)
        self.Label2.configure(background="#d9d9d9")
        self.Label2.configure(disabledforeground="#a3a3a3")
        self.Label2.configure(foreground="#000000")
        self.Label2.configure(text='''Enter an appropedia url below''')

if __name__ == '__main__':
    vp_start_gui()





